### PR TITLE
Add base documentation

### LINKED ISSUE #9: Add improved documentation for the existing codebase
### **Subtasks**

- [x] Add improved documentation for curvature/functions.jl
- [x] Add improved documentation for baselaplace.jl
- [x] Add improved documentation for laplace.jl
- [x] Add improved documentation for plotting.jl
- [x] Add improved documentation for utils.jl

### PR DESCRIPTION
**Pull request by Andrei32Ionescu**
Closes #9

### CODE REVIEW

File: src/baselaplace.jl [hunk 1]
```diff
@@ -19,8 +20,8 @@ Retrieves the desired (sub)set of model parameters and stores them in a list.
 
 ```julia-repl
 using Flux, LaplaceRedux
-nn = Chain(Dense(2,1))
-la = Laplace(nn)
+nn = Chain(Dense(2,1))                                      # defines a neural network with one hidden layer that takes a two-dimensional input 
+la = Laplace(nn)                                            # and produces a one-dimensional output
 LaplaceRedux.get_params(la)
```
Comment (severinbratus) [inline src/baselaplace.jl:25]: IMO having the comments on separate lines would make more sense, since you sometimes do not have a correspondence between the comment and the line

Comment (severinbratus) [reply src/baselaplace.jl:25]: This is a totally minor issue, so feel free to ignore

File: src/baselaplace.jl [hunk 2]
```diff
@@ -30,11 +31,11 @@ function get_params(la::BaseLaplace)
...
-        params = [θ for θ in params]                         # get all parameters and constants in logitbinarycrossentropy
+        params = [θ for θ in params]                                          # get all parameters and constants in logitbinarycrossentropy
     elseif la.subset_of_weights == :last_layer
-        params = [params[n_elements - 1], params[n_elements]]  # only get last parameters and constants
-    end
-    return params
+        params = [params[n_elements - 1], params[n_elements]]                   # only get last layer parameters and constants
+    end                                                                      # params[n_elements-1] is the weight matrix of the last layer
+    return params                                                            # params[n_elements] is the bias vector of the last layer
 end
```
Comment (severinbratus) [inline src/baselaplace.jl:39]: So the structure of `params` is like `[weights, biases, weights, biases, ...]`?

File: src/baselaplace.jl [hunk 3]
```diff
@@ -95,14 +96,14 @@ _init_H(la::BaseLaplace) = zeros(la.n_params, la.n_params)
...
 function _weight_penalty(la::BaseLaplace)
-    μ = la.μ    # MAP
-    μ₀ = la.μ₀  # prior
+    μ = la.μ                                                                 # MAP
+    μ₀ = la.μ₀                                                               # prior
     Δ = μ .- μ₀
-    return Δ'la.P₀ * Δ
-end
+    return Δ'la.P₀ * Δ                                                         # measure of how far the MAP estimate deviates from the prior mean
+end                                                                          # used to control the degree of regularization applied to the mode
```
Comment (severinbratus) [inline src/baselaplace.jl:106]: Here mode is MAP, right?

File: src/curvature/functions.jl [hunk 4]
```diff
@@ -102,14 +102,14 @@ end
 Compute the full empirical Fisher.
 """
 function full(curvature::EmpiricalFisher, d::Tuple)
-    x, y = d
+    x, y = d                                                                 # where x contains the observation's features and y represents the target value
 
     loss = curvature.factor * curvature.loss_fun(x, y)
     𝐠 = gradients(curvature, x, y)
-    𝐠 = reduce(vcat, [vec(𝐠[θ]) for θ in curvature.params])
+    𝐠 = reduce(vcat, [vec(𝐠[θ]) for θ in curvature.params])                    # concatenates the gradients into a vector
```
Comment (severinbratus) [inline src/curvature/functions.jl:109]: Theta here is an index, right? If so, I think `i` would be a better var name

File: src/laplace.jl [hunk 5]
```diff
@@ -10,13 +10,13 @@ mutable struct Laplace <: BaseLaplace
...
     hessian_structure::Symbol
     curvature::Union{Curvature.CurvatureInterface,Nothing}
-    σ::Real
-    μ₀::Real
-    μ::AbstractVector
-    P₀::Union{AbstractMatrix,UniformScaling}
-    H::Union{AbstractArray,Nothing}
-    P::Union{AbstractArray,Nothing}
-    Σ::Union{AbstractArray,Nothing}
+    σ::Real                                                                  # prior noise/uncertainty
```
Comment (severinbratus) [inline src/laplace.jl:13]: More precisely the standard deviation in the (Gaussian) prior

File: src/laplace.jl [hunk 6]
```diff
@@ -31,7 +31,7 @@ using Parameters
     backend::Symbol = :EmpiricalFisher
     σ::Real = 1.0
     μ₀::Real = 0.0
-    λ::Real = 1.0
+    λ::Real = 1.0                                                              # regularization parameter
```
Comment (severinbratus) [inline src/laplace.jl:34]: In a sidenote in https://www.paltmeyer.com/blog/posts/bayesian-logit/ PA equates `lambda == 1 / sigma^2`. Is this true in our case as well?

File: src/curvature/functions.jl [hunk 7]
```diff
@@ -16,9 +16,9 @@ function jacobians(curvature::CurvatureInterface, X::AbstractArray)
     # Output:
     ŷ = nn(X)
     # Jacobian:
-    𝐉 = jacobian(() -> nn(X), Flux.params(nn))
-    𝐉 = permutedims(reduce(hcat, [𝐉[θ] for θ in curvature.params]))
-    return 𝐉, ŷ
+    𝐉 = jacobian(() -> nn(X), Flux.params(nn))                                # differentiates f with regards to the model parameters
```
Comment (severinbratus) [inline src/curvature/functions.jl:19]: How does this work? We pass a function of zero args that returns the NN output for the input? But then if it was always a const value being returned, then the Jacobian would be zero. So how does Zygote compute this? By permuting parameters by an epsilon-value, and then seeing how `nn(X)` changes?

```
jacobian(f, args...) -> Tuple

For each array a ∈ args this returns a matrix with Ja[k,i] = ∂y[k]/∂a[i] where y = f(args...) is usually a vector. Arrays of higher dimension are treated like vec(a), or vec(y) for output.
```

File: src/baselaplace.jl [hunk 8]
```diff
@@ -95,14 +96,14 @@ _init_H(la::BaseLaplace) = zeros(la.n_params, la.n_params)
...
 
 """
     log_marginal_likelihood(la::BaseLaplace; P₀::Union{Nothing,UniformScaling}=nothing, σ::Union{Nothing, Real}=nothing)
```
Comment (severinbratus) [inline src/baselaplace.jl:109]: Can you comment on this as well? Where and for what is this used, what does it represent?

File: src/laplace.jl [hunk 9]
```diff
@@ -249,9 +249,9 @@ function optimize_prior!(
 )
 
     # Setup:
-    logP₀ = isnothing(λinit) ? log.(unique(diag(la.P₀))) : log.([λinit])   # prior precision (scalar)
-    logσ = isnothing(σinit) ? log.([la.σ]) : log.([σinit])                 # noise (scalar)
-    opt = Adam(lr)
+    logP₀ = isnothing(λinit) ? log.(unique(diag(la.P₀))) : log.([λinit])     # prior precision (scalar)
+    logσ = isnothing(σinit) ? log.([la.σ]) : log.([σinit])                   # noise (scalar)
+    opt = Adam(lr)                                                           # Adam is a stochastic gradient descent (SGD) optimization algorithm, using lr as learning rate
```
Comment (severinbratus) [inline src/laplace.jl:254]: On our discussion yesterday: Adam is applicable to mini-batch GD

File: src/curvature/functions.jl (hunk 7 above)
Comment (MarkArdman) [reply src/curvature/functions.jl:19]: I think this has less to do with us documenting it and more with understanding how Zygote works. Zygote is still able to compute the Jacobian by analyzing the computation graph of the neural network and its parameters even though the function passed to it has no arguments.

File: src/baselaplace.jl (hunk 2 above)
Comment (Andrei32Ionescu) [reply src/baselaplace.jl:39]: As I understand it, yes.

**Review by MarkArdman — APPROVED** [review]
Overall I don't think it gets much better than this, maybe the "main" exported functions like fit! and predict could benefit from more detailed documentation which explain what exactly and in what form the function returns, but that is not strictly necessary

File: src/laplace.jl (hunk 6 above)
Comment (Andrei32Ionescu) [reply src/laplace.jl:34]: I believe not, but I am unsure, we can check with PA next week. One of the reasons why is that they can be instantiated with different values in the _LaplaceParams_ constructor, which wouldn't be the case if there was an equivalence relation between them. If there is indeed such a relation between them, we should change the constructor to reflect that.

File: src/baselaplace.jl (hunk 1 above)
Comment (Andrei32Ionescu) [reply src/baselaplace.jl:26]: Fixed.

File: src/baselaplace.jl [hunk 10]
```diff
@@ -81,7 +82,7 @@ end
 """
     _H_factor(la::BaseLaplace)
 
-
+Returns the factor σ⁻², where σ is used in the centered Gaussian prior p(θ) = N(θ;0,σ²I)
```
[the remark this answers is not in the document]
Comment (Andrei32Ionescu) [reply src/baselaplace.jl]: Fixed.

File: src/curvature/functions.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (Andrei32Ionescu) [reply src/curvature/functions.jl]: Fixed.

Comment (Andrei32Ionescu) [reply src/curvature/functions.jl]: Agree.

File: src/laplace.jl (hunk 5 above)
Comment (Andrei32Ionescu) [reply src/laplace.jl]: Fixed.

File: src/curvature/functions.jl (hunk 7 above)
Comment (Andrei32Ionescu) [reply src/curvature/functions.jl]: I added a bit of bonus explanation, but most likely we will update it when we fully understand how Zygote works.

File: src/baselaplace.jl (hunk 8 above)
Comment (Andrei32Ionescu) [reply src/baselaplace.jl:113]: Done!

File: src/laplace.jl (hunk 9 above)
Comment (Andrei32Ionescu) [reply src/laplace.jl]: Yes, fixed it.

File: src/laplace.jl [hunk 11]
```diff
@@ -251,7 +252,7 @@ function optimize_prior!(
     # Setup:
     logP₀ = isnothing(λinit) ? log.(unique(diag(la.P₀))) : log.([λinit])     # prior precision (scalar)
     logσ = isnothing(σinit) ? log.([la.σ]) : log.([σinit])                   # noise (scalar)
-    opt = Adam(lr)                                                           # Adam is a stochastic gradient descent (SGD) optimization algorithm, using lr as learning rate
+    opt = Adam(lr)                                                           # Adam is gradient descent (GD) optimization algorithm, using lr as learning rate
```
Comment (severinbratus) [inline src/laplace.jl:255]: Oh, I believe minibatch gradient descent is still considered to be stochastic -- but nevermind, my comment was meant simply as a comment

**Review by severinbratus — APPROVED** [review]
This served as a good intro to the repository, I believe the comments are enough to help us to get some insights on the code.

**Review by navimakarov — APPROVED** [review]
Looks good!
